Refactor specs related to Active Job

monfresh opened the pull request:
**Why**:
- To make them more robust and less prone to intermittent failures

**How**:
- Instead of relying on a gem to check if a job was enqueued, check
that a specific method was called on an object, since those objects are
already unit tested separately
- Replace feature specs with controller and view specs
- Make sure Active Job queues are cleared **before** specs that
interact with them, as opposed to **after**.

For example, when a user signs in, they get sent an OTP via the
`send_two_factor_authentication_code` method. That method then
instantiates the `UserOtpSender` class, which contains logic that
determines which devices the OTP should be sent to. That logic is unit
tested separately in `otp_sender_spec`. So, instead of writing various
slow feature tests for all the different scenarios while signing in, we
just need to write one `SessionsController` test that verifies that
when a user signs in, the `send_two_factor_authentication_code` is
called on the current user.

monfresh commented:
Could I get someone to look at this please? @jgrevich @amos?

monfresh commented:
Ping. Could someone please take a look? This is fairly straightforward and eliminates the flickering specs from Travis.

amoose commented:
👓

amoose commented on spec/mailers/email_second_factor_mailer_spec.rb:
```diff
@@ -16,5 +16,10 @@
+
+    it 'includes a link to customer service in the email' do
+      expect(mail.body).
+        to include 'at <a href="https://upaya.18f.gov/contact">'
```
it's probably worth adding url to locale as well

monfresh replied:
Agreed, but that's beyond the scope of this PR. Will open a new issue.

amoose replied:
👌

amoose commented on spec/services/twilio_service_spec.rb:
```diff
@@ -50,28 +50,28 @@
       expect(FeatureManagement).to receive(:pt_mode?).at_least(:once).and_return(false)
       SmsSenderOtpJob.perform_now(user)
 
       expect(messages.first.from).to eq '+19999999999'
```
pull from ~~application.yml~~ secrets.yml and remove unused keys from application.yml.example

monfresh replied:
Good point, but I think that's beyond the scope of this PR. The twilio keys in application.yml are not used in practice because we have the correct keys in secrets.yml, but we have the `self.singular_account` method in `TwilioService` that falls back to the keys in application.yml if the ones in secrets.yml are not present. So, they are actually used in a sense.

We can open an issue to address this later. Ideally, we would do something like the required keys initializer that Figaro provides so that the app won't run if keys are missing.

amoose commented on spec/support/controller_helper.rb:
```diff
@@ -17,6 +17,12 @@ def sign_in_as_tech_user
+    allow(warden).to receive(:authenticated?).with(:user).and_return(true)
+    allow(controller).to receive(:current_user).and_return(user)
+    warden.session(:user)[TwoFactorAuthentication::NEED_AUTHENTICATION] = true
+  end
```
rename to `sign_in_bypassing_2fa` or something more descriptive

amoose commented:
Awesome spec improvements 🍷

monfresh commented:
Rebased and ready to go. I renamed that controller helper to `sign_in_before_2fa`. What we are testing here is that when a user that is not yet fully authenticated visits the show action in `TwoFactorAuthenticationController`, the app renders the show template. I had to create a new helper for this because the default `sign_in` helper automatically fully authenticates the user. Let me know if the new name better describes what it does.

amoose commented on spec/features/two_factor_authentication/sign_in_spec.rb:
```diff
@@ -106,7 +106,7 @@
+      context 'user enters a valid number then changes their mind' do
         before do
           sign_up_with_and_set_password_for('test@example.com')
           @user = User.find_by_email('test@example.com')
```
we can replace instances of:
```
sign_up_with_and_set_password_for('test@example.com')
@user = User.find_by_email('test@example.com')
```
with:
`@user = sign_up_with_and_set_password_for`

amoose commented:
Another fine PR 🍷
